pkg/webhook/clusterset/validatingWebhook.go

```diff
@@ -0,0 +1,356 @@
+	if err != nil {
+		status.Allowed = false
+		status.Result = &metav1.Status{
+			Status: metav1.StatusFailure, Code: http.StatusBadRequest, Reason: metav1.StatusReasonBadRequest,
+			Message: err.Error(),
+		}
+		return status
+	}
+
+	switch request.Resource {
+	case clusterDeploymentsGVR:
+		// ignore clusterDeployment created by clusterClaim
+		claimName, _, err := unstructured.NestedString(obj.Object, "spec", "clusterPoolRef", "claimName")
+		if err != nil {
+			status.Allowed = false
+			status.Result = &metav1.Status{
+				Status: metav1.StatusFailure, Code: http.StatusBadRequest, Reason: metav1.StatusReasonBadRequest,
+				Message: err.Error(),
+			}
+			return status
+		}
+		if claimName != "" {
+			return status
+		}
+	}
+
+	labels := obj.GetLabels()
+	clusterSetName := ""
+	if len(labels) > 0 {
+		clusterSetName = labels[clusterSetLabel]
+	}
+
+	if clusterSetName == "" {
+		status.Allowed = false
+		status.Result = &metav1.Status{
+			Status: metav1.StatusFailure, Code: http.StatusForbidden, Reason: metav1.StatusReasonForbidden,
+			Message: fmt.Sprintf("user %q is forbidden to create a resource without ManagedClusterSet label", request.UserInfo.Username),
+		}
+		return status
+	}
+
+	return a.allowSetClusterSetLabel(request.UserInfo, "", clusterSetName)
+}
+
+func (a *AdmissionHandler) validateUpdateRequest(request *v1.AdmissionRequest) *v1.AdmissionResponse {
+	status := &v1.AdmissionResponse{
+		Allowed: true,
+	}
+
+	oldObj := unstructured.Unstructured{}
+	err := oldObj.UnmarshalJSON(request.OldObject.Raw)
+	if err != nil {
+		status.Allowed = false
+		status.Result = &metav1.Status{
+			Status: metav1.StatusFailure, Code: http.StatusBadRequest, Reason: metav1.StatusReasonBadRequest,
+			Message: err.Error(),
+		}
+		return status
+	}
+
+	newObj := unstructured.Unstructured{}
+	err = newObj.UnmarshalJSON(request.Object.Raw)
+	if err != nil {
+		status.Allowed = false
+		status.Result = &metav1.Status{
+			Status: metav1.StatusFailure, Code: http.StatusBadRequest, Reason: metav1.StatusReasonBadRequest,
+			Message: err.Error(),
+		}
+		return status
+	}
+
+	oldLabels := oldObj.GetLabels()
+	newLabels := newObj.GetLabels()
+	originalClusterSetName := ""
+	currentClusterSetName := ""
+	if len(oldLabels) > 0 {
+		originalClusterSetName = oldLabels[clusterSetLabel]
+	}
+	if len(newLabels) > 0 {
+		currentClusterSetName = newLabels[clusterSetLabel]
+	}
+
+	// allow if clusterset label is not updated
+	if originalClusterSetName == currentClusterSetName {
+		return status
+	}
+
+	if currentClusterSetName == "" {
+		status.Allowed = false
+		status.Result = &metav1.Status{
+			Status: metav1.StatusFailure, Code: http.StatusForbidden, Reason: metav1.StatusReasonForbidden,
+			Message: fmt.Sprintf("user %q is forbidden to update a resource without ManagedClusterSet label", request.UserInfo.Username),
+		}
+		return status
+	}
+
+	return a.allowSetClusterSetLabel(request.UserInfo, originalClusterSetName, currentClusterSetName)
```

Comment:
To make it simple I think you only need a.allowSetClusterSetLabel(request.UserInfo, originalClusterSetName),

so if user did not set  clustersetname, you can still use this to func to check. For update operation, run the func twice with old and new clusterset  name.

Reply:
done